baybe/utils/augmentation.py

```diff
@@ -0,0 +1,242 @@
+
+    *   Original
+
+        +---+---+---+---+
+        | A | B | C | D |
+        +===+===+===+===+
+        | a | b | x | y |
+        +---+---+---+---+
+        | b | a | x | z |
+        +---+---+---+---+
+
+    *   Result with ``columns = ["A", "B"]``
+
+        +---+---+---+---+
+        | A | B | C | D |
+        +===+===+===+===+
+        | a | b | x | y |
+        +---+---+---+---+
+        | b | a | x | z |
+        +---+---+---+---+
+        | b | a | x | y |
+        +---+---+---+---+
+        | a | b | x | z |
+        +---+---+---+---+
+
+    *   Result with ``columns = ["A", "B"]``, ``dependents = ["C", "D"]``
+
+        +---+---+---+---+
+        | A | B | C | D |
+        +===+===+===+===+
+        | a | b | x | y |
+        +---+---+---+---+
+        | b | a | x | z |
+        +---+---+---+---+
+        | b | a | y | x |
+        +---+---+---+---+
+        | a | b | z | x |
+        +---+---+---+---+
+
+    Args:
+        df: The dataframe that should be augmented.
+        columns: The permutation invariant columns.
+        dependents: Columns that are connected to `columns` and should be permuted in
+            the same manner.
+
+    Returns:
+        The augmented dataframe containing the original one.
+
+    Raises:
+        ValueError: If `dependents` has length incompatible with `columns`.
+    """
+    dependents = dependents or []
+    new_rows: list[pd.DataFrame] = []
+
+    if dependents and len(columns) != len(dependents):
+        raise ValueError(
+            "When augmenting permutation invariance with dependent columns, there must "
+            "be exactly the same amount of 'dependents' as there are 'columns'."
+        )
+
+    for _, row in df.iterrows():
+        # Extract the values from the specified columns
+        original_values = row[columns].tolist()  # type: ignore[call-overload]
+        dependent_values = row[dependents].tolist() if dependents else None  # type: ignore[call-overload]
+
+        # Generate all permutations of these values
+        column_perms = list(permutations(original_values))
+        dependent_perms = (
+            list(permutations(dependent_values)) if dependent_values else None
+        )
+
+        # For each permutation, create a new row if it's not already in the dataframe
+        for k, perm in enumerate(column_perms):
+            # Create a new row dictionary with the permuted values
+            new_row = pd.DataFrame([row])
+            new_row[columns] = perm
+            if dependent_perms:
+                new_row[dependents] = dependent_perms[k]
+
+            if not _row_in_df(new_row, df):
+                new_rows.append(new_row)
+
+    augmented_df = pd.concat([df] + new_rows)
+
+    return augmented_df
+
+
+def df_apply_dependency_augmentation(
+    df: pd.DataFrame,
+    causing: tuple[str, Sequence],
+    affected: Sequence[tuple[str, Sequence]],
+) -> pd.DataFrame:
```

Comment:
as you say, this scenario is not currently possible due to the design of the dependency constraint
yes it could potentially change in the future - that is the point to expand this utility, imo not now